File: src/main/java/io/github/saneea/FeatureProvider.java
package io.github.saneea;

import java.util.Set;

import io.github.saneea.feature.multi.FeatureInfo;

public interface FeatureProvider {
	Set<String> featuresNames();

	Feature createFeature(String featureName) throws Exception;

	default FeatureInfo featureInfo(FeatureContext context) {
		try {
			Feature feature = createFeature(context.getFeatureName());
			return new FeatureInfo(feature, feature.meta(context).description().brief());
		} catch (Exception e) {
			return new FeatureInfo(null, e.toString());
		}
	}
}

File: src/main/java/io/github/saneea/dvh/FeatureProvider.kt
package io.github.saneea.dvh

import io.github.saneea.dvh.feature.multi.FeatureInfo

interface FeatureProvider {

    fun featuresNames(): Set<String>

    fun createFeature(featureName: String): Feature?

    fun featureInfo(context: FeatureContext): FeatureInfo {
        return try {
            val feature = createFeature(context.featureName, context)
            FeatureInfo(feature, feature!!.meta.description.brief)
        } catch (e: Exception) {
            FeatureInfo(null, e.toString())
        }
    }
}

fun FeatureProvider.createFeature(featureName: String, context: FeatureContext) =
    this.createFeature(featureName).also { it?.context = context }